File: Week 3/Python Codes/MiddleOfLinkedList.py
'''
class node:
    def __init__(data):
        self.data = data
        self.next = None
'''


# Method 1: Iterative Approach
def findMid(head):
    total, temp = 0, head
    if temp is None:
        return -1
    while temp is not None:
        temp = temp.next
        total += 1
    
    count, temp = 0, head
    while count <= total//2:
        if count == total//2:
            return temp.data
        count += 1
        temp = temp.next



# Method 2: Two Pointer Approach
def findMid(head):
    if head is None:
        return -1
    slowPtr, fastPtr = head, head
    if head is not None:
        while (fastPtr is not None and fastPtr.next is not None):
            slowPtr = slowPtr.next
            fastPtr = fastPtr.next.next
    return slowPtr.data

File: Week 3/C++ Codes/MiddleOfLinkedList.cpp
#include<bits/stdc++.h>
using namespace std;

// Link list Node 
struct Node {
    int data;
    Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
    
};

// Method - 1: Iterative Approach
int getMiddle(Node *head)
{
    int length = 0, counter = 0;
    struct Node* temp = head;
    if (head == NULL)
        return -1;
    while (temp != NULL) {
        temp = temp -> next;
        length += 1;
    }
    temp = head, counter = 0;
    while (counter <= length / 2) {
        if (counter == (length/2))
            return temp->data;
        temp = temp->next;
        counter += 1;
    }    
}

// Method 2 - Two Pointer Approach
int getMiddle(Node *head)
{
    struct Node *fastPtr = head;
    struct Node *slowPtr = head;
    if (head == NULL)
        return -1;
    
    if (head != NULL) {
        while (fastPtr != NULL && fastPtr->next != NULL) {
            fastPtr = fastPtr->next->next;
            slowPtr = slowPtr->next;
        }
    }
    return slowPtr->data;
}
